### PR TITLE
Fix a regression in 2.13.2 where a ``RunTimeError`` could be raised unexpectedly

### LINKED ISSUE #1958: v2.13.x regression: Crash when inspecting `PyQt5.QtWidgets` due to `RuntimeError` during `hasattr`
### Steps to reproduce

Install PyQt5, run `pylint --extension-pkg-whitelist=PyQt5 x.py` over a file containing `from PyQt5 import QtWidgets`

### Current behavior

With astroid 2.12.13 and pylint 2.15.10, this works fine. With astroid 2.13.2, this happens:

```pytb
Exception on node <ImportFrom l.1 at 0x7fc5a3c47d00> in file '/home/florian/tmp/pylintbug/x.py'
Traceback (most recent call last):
  File "/home/florian/tmp/pylintbug/.venv/lib/python3.10/site-packages/pylint/utils/ast_walker.py", line 90, in walk
    callback(astroid)
  File "/home/florian/tmp/pylintbug/.venv/lib/python3.10/site-packages/pylint/checkers/variables.py", line 1726, in visit_importfrom
    self._check_module_attrs(node, module, name.split("."))
  File "/home/florian/tmp/pylintbug/.venv/lib/python3.10/site-packages/pylint/checkers/variables.py", line 2701, in _check_module_attrs
    module = next(module.getattr(name)[0].infer())
  File "/home/florian/tmp/pylintbug/.venv/lib/python3.10/site-packages/astroid/nodes/scoped_nodes/scoped_nodes.py", line 412, in getattr
    result = [self.import_module(name, relative_only=True)]
  File "/home/florian/tmp/pylintbug/.venv/lib/python3.10/site-packages/astroid/nodes/scoped_nodes/scoped_nodes.py", line 527, in import_module
    return AstroidManager().ast_from_module_name(
  File "/home/florian/tmp/pylintbug/.venv/lib/python3.10/site-packages/astroid/manager.py", line 205, in ast_from_module_name
    return self.ast_from_module(named_module, modname)
  File "/home/florian/tmp/pylintbug/.venv/lib/python3.10/site-packages/astroid/manager.py", line 312, in ast_from_module
    return AstroidBuilder(self).module_build(module, modname)
  File "/home/florian/tmp/pylintbug/.venv/lib/python3.10/site-packages/astroid/builder.py", line 101, in module_build
    node = self.inspect_build(module, modname=modname, path=path)
  File "/home/florian/tmp/pylintbug/.venv/lib/python3.10/site-packages/astroid/raw_building.py", line 366, in inspect_build
    self.object_build(node, module)
  File "/home/florian/tmp/pylintbug/.venv/lib/python3.10/site-packages/astroid/raw_building.py", line 422, in object_build
    elif hasattr(member, "__all__"):
RuntimeError: wrapped C/C++ object of type QApplication has been deleted
x.py:1:0: F0002: x.py: Fatal error while checking 'x.py'. Please open an issue in our bug tracker so we address this. There is a pre-filled template that you can use in '/home/florian/.cache/pylint/pylint-crash-2023-01-10-11-06-17.txt'. (astroid-error)
```

It looks like it happens when `member` is `QtWidgets.qApp`, which is a kind of "magic" object referring to the QApplication singleton. Since none exists, it looks like PyQt doesn't like trying to access an attribute on that.

Bisected to:

- #1885 

It looks like 974f26f75eb3eccb4bcd8ea143901baf60a685ff is the exact culprit.

cc @nickdrozd 

(took the freedom to add appropriate labels already, hope that's fine)

### PR DESCRIPTION
**Pull request by Pierre-Sassoulas**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |

## Description

Co-authored-by: Florian Bruhin <me@the-compiler.org>

Closes #1958

### CODE REVIEW

**Comment by The-Compiler** [comment]
I've tried various ways to write a test for this:

```diff
diff --git c/tests/testdata/python3/data/fake_module_with_broken_getattr.py i/tests/testdata/python3/data/fake_module_with_broken_getattr.py
new file mode 100644
index 00000000..e860928e
--- /dev/null
+++ i/tests/testdata/python3/data/fake_module_with_broken_getattr.py
@@ -0,0 +1,7 @@
+class Broken:
+
+    def __getattr__(self, name):
+        raise Exception("boom")
+
+
+broken = Broken()
diff --git c/tests/unittest_raw_building.py i/tests/unittest_raw_building.py
index 59bd6ce1..e00c6760 100644
--- c/tests/unittest_raw_building.py
+++ i/tests/unittest_raw_building.py
@@ -8,8 +8,10 @@
 import _io
 import pytest
 
-# A fake module to simulate pandas in unittest below
+# A fake module to simulate issues in unittest below
 import tests.testdata.python3.data.fake_module_with_warnings as fm
+import tests.testdata.python3.data.fake_module_with_broken_getattr as fm_getattr
+
 from astroid.builder import AstroidBuilder
 from astroid.const import IS_PYPY
 from astroid.raw_building import (
@@ -102,6 +104,20 @@ def test_build_function_deepinspect_deprecation(self) -> None:
         # This should not raise an exception
         AstroidBuilder().module_build(m, "test")
 
+    def test_module_object_with_broken_getattr(self) -> None:
+        # Tests https://github.com/PyCQA/astroid/issues/1958
+        # When astroid deep inspection of modules raises
+        # errors when using hasattr().
+
+        # Create a mock module to simulate a broken module
+        m = types.ModuleType("test")
+
+        # Attach a mock of pandas with the same behavior
+        m.submodule = fm_getattr
+
+        # This should not raise an exception
+        AstroidBuilder().module_build(m, "test")
+
 
 if __name__ == "__main__":
     unittest.main()
diff --git c/tests/unittest_regrtest.py i/tests/unittest_regrtest.py
index 806e6fe0..2d2db616 100644
--- c/tests/unittest_regrtest.py
+++ i/tests/unittest_regrtest.py
@@ -66,6 +66,12 @@ def test_living_property(self) -> None:
         builder._module = sys.modules[__name__]
         builder.object_build(build_module("module_name", ""), Whatever)
 
+    def test_broken_hasattr(self) -> None:
+        builder = AstroidBuilder()
+        builder._done = {}
+        builder._module = sys.modules[__name__]
+        build_module("module_name", Broken)
+
     @unittest.skipIf(not HAS_NUMPY, "Needs numpy")
     def test_numpy_crash(self):
         """test don't crash on numpy"""
@@ -343,6 +349,12 @@ class Whatever:
     a = property(lambda x: x, lambda x: x)  # type: ignore[misc]
 
 
+class Broken:
+
+    def __getattr__(self, attr):
+        raise Exception("boom")
+
+
 def test_ancestor_looking_up_redefined_function() -> None:
     code = """
     class Foo:

```

but can't seem to get it to work....

But I believe this will trigger with *any* exception (other than `AttributeError`) in a `__getattr__`, as long as that's in a C extension module where pylint actually infers the real module content dynamically.

**Comment by Pierre-Sassoulas** [comment]
It seems we'd need to create a c extensions to test on and compile it, right ?

**Comment by The-Compiler** [comment]
> It seems we'd need to create a c extensions to test on and compile it, right ?

For an integration test, sure - I was hoping to be able to call the affected `inspect_build` / `object_build` code directly from an unit test. Looking at the above code again, it looks like I'm only calling `AstroidBuilder.module_build()` which then indeed (I think) calls `file_build` instead.

But this seems to work:


```diff
commit 211e71597a38fdbf68e38b6ab2ec4da901a54d4e
Author: Florian Bruhin <me@the-compiler.org>
Date:   Tue Jan 10 18:48:34 2023 +0100

    Add a unit test for broken __getattr__ in extension modules

diff --git a/tests/testdata/python3/data/fake_module_with_broken_getattr.py b/tests/testdata/python3/data/fake_module_with_broken_getattr.py
new file mode 100644
index 00000000..e860928e
--- /dev/null
+++ b/tests/testdata/python3/data/fake_module_with_broken_getattr.py
@@ -0,0 +1,7 @@
+class Broken:
+
+    def __getattr__(self, name):
+        raise Exception("boom")
+
+
+broken = Broken()
diff --git a/tests/unittest_raw_building.py b/tests/unittest_raw_building.py
index 59bd6ce1..983736e4 100644
--- a/tests/unittest_raw_building.py
+++ b/tests/unittest_raw_building.py
@@ -8,8 +8,10 @@
 import _io
 import pytest
 
-# A fake module to simulate pandas in unittest below
+# A fake module to simulate issues in unittest below
 import tests.testdata.python3.data.fake_module_with_warnings as fm
+import tests.testdata.python3.data.fake_module_with_broken_getattr as fm_getattr
+
 from astroid.builder import AstroidBuilder
 from astroid.const import IS_PYPY
 from astroid.raw_building import (
@@ -102,6 +104,14 @@ def test_build_function_deepinspect_deprecation(self) -> None:
         # This should not raise an exception
         AstroidBuilder().module_build(m, "test")
 
+    def test_module_object_with_broken_getattr(self) -> None:
+        # Tests https://github.com/PyCQA/astroid/issues/1958
+        # When astroid deep inspection of modules raises
+        # errors when using hasattr().
+
+        # This should not raise an exception
+        AstroidBuilder().inspect_build(fm_getattr, "test")
+
 
 if __name__ == "__main__":
     unittest.main()

```

If you want to apply that, you can pipe it to `git am` - or I can push it into the branch, if you prefer.

**Comment by Pierre-Sassoulas** [comment]
I had to educate myself about ``git am`` first. That's pretty nice, better than ``git apply``, thank you.  Thank you also for the patch !

**Comment by Pierre-Sassoulas** [comment]
The test is properly failing on main, let's merge.

File: tests/unittest_raw_building.py [hunk 1]
```diff
@@ -1,3 +1,8 @@
+"""
+'tests.testdata.python3.data.fake_module_with_warnings' is a fake module to
+simulate issues in unittest below
+"""
```
Comment (Pierre-Sassoulas) [inline tests/unittest_raw_building.py:4]:
```suggestion
"""
'tests.testdata.python3.data.fake_module_with_warnings' and
'tests.testdata.python3.data.fake_module_with_warnings' are fake modules
to simulate issues in unittest below
"""
```

**Pull request merged**